tools/wptrunner/wptrunner/wptrunner.py

```diff
@@ -235,143 +360,83 @@ def run_tests(config, test_paths, product, **kwargs):
             recording.set(["startup", "ensure_environment"])
             try:
                 test_environment.ensure_started()
+                start_time = datetime.now()
             except env.TestEnvironmentError as e:
                 logger.critical("Error starting test environment: %s" % e)
                 raise
 
             recording.set(["startup"])
 
+            max_time = None
+            if "repeat_max_time" in kwargs:
+                max_time = timedelta(minutes=kwargs["repeat_max_time"])
+
             repeat = kwargs["repeat"]
-            repeat_count = 0
             repeat_until_unexpected = kwargs["repeat_until_unexpected"]
 
-            while repeat_count < repeat or repeat_until_unexpected:
-                repeat_count += 1
+            # keep track of longest time taken to complete a
+            # test suite iteration so that the runs can be stopped
+            # to avoid a possible TC timeout.
+            longest_iteration_time = timedelta()
+            # keep track if we break the loop to avoid timeout.
+            iteration_timeout = False
+
+            while counts["repeat"] < repeat or repeat_until_unexpected:
+                # if the next repeat run could cause the TC timeout to be
+                # reached, stop now and use the test results we have.
+                estimate = datetime.now() + longest_iteration_time
+                if not repeat_until_unexpected and max_time \
+                        and estimate >= start_time + max_time:
+                    iteration_timeout = True
+                    logger.info(
```

Comment:
Although there's one existing example of the "a problem occured!" style of error mssage, I think I'd prefer to avoid that if possible. Also at risk of being needlessly pedantic, the formatting here which breaks things into three lines when they could be two seems needlessly wasteful of vertical space.

Reply:
I will remove this output and change the formatting 👍